proposals/NNNN-ComparisonReform.md

```diff
@@ -0,0 +1,531 @@
+work but floats don't is a poor state of affairs. One should be able to sort an
+array of floats and use floats as keys in data structures, even if the latter
+is difficult to do correctly.
+
+### PartialComparable
+
+PartialComparable would essentially just be Comparable without any stated
+ordering requirements, that Comparable extends to provide ordering
+requirements. This would be a protocol that standard IEEE comparison could
+satisfy, but in the absence of total ordering requirements, PartialComparable
+is effectively useless. Either everyone would consume PartialComparable (to
+accept floats) or Comparable (to have reasonable behaviour).
+
+The Rust community adopted this strategy to little benefit. The Rust libs team
+has frequently considered removing the distinction, but hasn't because doing it
+backwards compatibly would be complicated. Also because merging the two would
+just lead to the problems Swift has today.
+
+### Different Names For `compare` and `ComparisonResult`
+
+A few different variants for `ComparisonResult` and its variants were considered:
+
+* Dropping the `ordered` part of `ComparisonResult`'s cases e.g. `.ascending`
+* Naming of `ComparisonResult` as `SortOrder`
+* `enum Ordering { case less, equal, greater }` ([as used by Rust](https://doc.rust-lang.org/std/cmp/enum.Ordering.html))
+* Case values of `inOrder`, `same`, `outOfOrder`
+
+The choice of case names is non-trivial because the enum shows up in different
+contexts where different names makes more sense. Effectively, one needs to keep
+in mind that the "default" sort order is ascending to map between the concept
+of "before" and "less".
+
+The before/after naming to provide the most intuitive model for custom sorts --
+referring to `ascending` or `less` is confusing when trying to implement a
+descending ordering. Similarly the inOrder/outOfOrder naming was too indirect
+-- it's more natural to just say where to put the element. If the enum should
+focus on the sorting case, calling it `SortOrder` would help to emphasize this
+fact.
+
+This proposal elects to leave the existing Foundation name in-place. The
+primary motivation for this is that use of the `compare` function will be
+relatively rare. It is expected that in most cases users will continue to make
+use of `==` or `<`, returning boolean values (the main exception to this will
+be in use of the parameterized `String` comparisons). As such, the source
+compatibility consequences of introducing naming changes to an existing type
+seems of insufficient benefit.
+
+The method `compare(_:)` does not fully comport with the API naming guidelines.
+However, it is firmly established with current usage in Objective-C APIs, will
+be fairly rarely seen/used (users will usually prefer `<`, `==` etc), and
+alternatives considered, for example `compared(to:)`, were not a significant
+improvement.
+
+### Add Overloads for `(T, T) -> ComparisonResult`
```

Comment:
I would strongly support replacing `(T, T) -> Bool` with `(T, T) -> ComparisonResult` variants—not simply supplementing. Here's why:

1) The `areInIncreasingOrder` variant is confusing at the call-site since the definition must be consulted to determine which order the comparison expects.
2) The `areInIncreasingOrder` implementation is very dirty when an `==` result is desired. This is especially bad if we expect other authors to mirror this API:
```swift
if !areInIncreasingOrder(a, b) && !areInIncreasingOrder(b, a) {
  // equal!
}
```
Not only is this unintuitive, but it is also less performant in many cases.